### PR TITLE
notebook of xdeepfm on movielens with libffmconverter

### PR DESCRIPTION
**Pull request by yexing99**
### Description

1. Added libffmconverter to convert movielens genres features into libffm format. 
2. Added notebook of xdeepfm running on movielens data 
3. This notebook would be further polished once libffmconverter is updated. The current notebook shows a **workaround** using libffmconverter to change different features within one field into libffm format

1. Provided notebook on how to use xdeepfm on movielens
2. Showed a workaround on libffmconverter in how to deal with multiple features within one field
 
### Related Issues

https://github.com/microsoft/recommenders/issues/446

### Checklist:

- [ ] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [ ] I have added tests covering my contributions.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: notebooks/00_quick_start/xdeepfm_movielens.ipynb [hunk 1]
```diff
@@ -0,0 +1,1541 @@
...
+       "      <td>19:2642:0</td>\n",
+       "      <td>20:2643:0</td>\n",
+       "      <td>21:2644:0</td>\n",
+       "    </tr>\n",
+       "  </tbody>\n",
+       "</table>\n",
+       "<p>5 rows × 22 columns</p>\n",
+       "</div>"
+      ],
+      "text/plain": [
```
Comment (yueguoguo) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: @yexing99 in the settings here, user, item and genres are treated as fields, and the features in each field should be counted in the dictionary. Not sure if it is right to restart the count from 1 from item field to genre field. 

Add @Leavingseason

File: notebooks/00_quick_start/xdeepfm_movielens.ipynb [hunk 2]
```diff
@@ -0,0 +1,1541 @@
...
+    "### Parameters"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Download data\n",
+    "xDeepFM uses the FFM format as data input: `<label> <field_id>:<feature_id>:<feature_value>`  \n",
+    "Each line represents an instance, `<label>` is a binary value with 1 meaning positive instance and 0 meaning negative instance. \n",
```
Comment (yueguoguo) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: It looks like the problem in this notebook is to predict ratings. Then the labels are numerical instead of binary values.

File: notebooks/00_quick_start/xdeepfm_movielens.ipynb [hunk 3]
```diff
@@ -0,0 +1,1541 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "step 20 , total_loss: 3.7089, data_loss: 3.7089\n",
```
Comment (yueguoguo) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: Looks like after epoch 6 the loss metric starts to vibrate around some values (e.g., 1.2). There might be something to do with the hyperparameters. Since this notebook is just a quickstart, so it would be fine not to provide codes for tuning, but maybe it is worth showing the best parameters and results from experimentation that has tuning.

File: reco_utils/recommender/deeprec/deeprec_utils.py [hunk 4]
```diff
@@ -389,6 +388,14 @@ def cal_metric(labels, preds, metrics):
             pred[pred < 0.5] = 0
             f1 = f1_score(np.asarray(labels), pred)
             res["f1"] = round(f1, 4)
+        elif metric == 'mae':
```
Comment (miguelgfierro) [inline reco_utils/recommender/deeprec/deeprec_utils.py]: we had a FIXME at the beginning of this function to make it explicit. We also have an issue to do the same in the LigthGBM notebook https://github.com/microsoft/recommenders/issues/762

do you think you could refactor this?

File: tests/resources/deeprec/movielens/network_xdeepFM.yaml [hunk 5]
```diff
@@ -0,0 +1,61 @@
+#data
```
Comment (miguelgfierro) [inline tests/resources/deeprec/movielens/network_xdeepFM.yaml]: is it possible to eliminate the yaml file and add the information directly to the notebook? 

To upload to github a yaml file is tricky because if you want to do experiments and change the inputs, we need to modify this file, which gets tracked. 

I had this same conversation with Jianxun and finally he download a yaml file from a blob in the notebook and then set some parameters directly. Is it possible to just set the parameters in the notebook?

File: reco_utils/recommender/deeprec/deeprec_utils.py [hunk 6]
```diff
@@ -389,6 +388,14 @@ def cal_metric(labels, preds, metrics):
...
+            res['mae'] = mean_absolute_error(np.asarray(labels), np.asarray(preds))
+        elif metric == 'rsquare':
+            res['rsquare'] = r2_score(np.asarray(labels), np.asarray(preds))
```
Comment (miguelgfierro) [inline reco_utils/recommender/deeprec/deeprec_utils.py]: I think np.asarray is not needed, if I'm not mistaken it is slower and the result is the same without it

File: notebooks/00_quick_start/xdeepfm_movielens.ipynb [hunk 7]
```diff
@@ -0,0 +1,1541 @@
+{
```
Comment (miguelgfierro) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: <p>in the other notebooks we are placing all the imports in the first cell</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/861/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (miguelgfierro) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: <p>this might break if the folders are not created, probably it is safer to do just `data_path = "."`</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/861/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (miguelgfierro) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: <p>what I was saying in the other comments is removing the ymal file and add here a dictionary of the parameters we want to add. I think it is easier for the user to understand. Maybe we can explain some of them</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/861/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

Comment (miguelgfierro) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: <p>is there a way to be less verbose?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/861/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

Comment (miguelgfierro) [inline notebooks/00_quick_start/xdeepfm_movielens.ipynb]: <p>apart from this content it would be good to test this notebook. For it you will need to activate the pm.record option and follow this instructions: <a href="https://github.com/microsoft/recommenders/tree/master/tests#how-to-create-tests-on-notebooks-with-papermill" target="_blank">https://github.com/microsoft/recommenders/tree/master/tests#how-to-create-tests-on-notebooks-with-papermill</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/861/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='29'/>

**Comment by atimesastudios** [comment]
Thank you for adding some more examples of how to setup and use xDeepFM. The current notebook example that uses synthetic data and criteo data takes it through training, evaluation and then produces outputs from the test file. It is not clear what those outputs represent and I'm not sure how to get top k recommendations out of this model. Depending if linear, FM, CIN, DNN etc. are selected in the hyperparameter file are we supposed to interpret the outputs differently? I've read Lian, Zhou et. al paper from 2018 and looked at the Leavingseason/xDeepFM repo and I must be missing a trick somewhere! I'm passing in ratings on a 1-5 scale and getting outputs from the test file that seem to be on a different scale. I hope that the movielens example will take it through to the point that we can deploy and productionize models using xDeepFM. If I can be of any assistance in helping out please let me know.

**Comment by yueguoguo** [comment]
> Thank you for adding some more examples of how to setup and use xDeepFM. The current notebook example that uses synthetic data and criteo data takes it through training, evaluation and then produces outputs from the test file. It is not clear what those outputs represent and I'm not sure how to get top k recommendations out of this model. Depending if linear, FM, CIN, DNN etc. are selected in the hyperparameter file are we supposed to interpret the outputs differently? I've read Lian, Zhou et. al paper from 2018 and looked at the Leavingseason/xDeepFM repo and I must be missing a trick somewhere! 

The output file if you specify in either yaml file or the input argument of the model object will write the prediction results on the testing data into the file. Then you need to manually read the file and append the results into the original testing data. Getting the top-k results will require an additional step to rank the scores. In this case, you may need to prepare the testing data as a cross-join of all users and items to make sure they are counted in the ranking. 

> I'm passing in ratings on a 1-5 scale and getting outputs from the test file that seem to be on a different scale. I hope that the movielens example will take it through to the point that we can deploy and productionize models using xDeepFM. If I can be of any assistance in helping out please let me know.

Did you set the task to be `regression`?
